packages/gatsby/src/redux/nodes.js

```diff
@@ -154,3 +154,94 @@ const addResolvedNodes = (typeName, arr) => {
 }
 
 exports.addResolvedNodes = addResolvedNodes
+
+// TODO: This local cache is "global" to the nodejs instance. It should be
+//       reset after each point where nodes are mutated. This is currently
+//       not the case. I don't think we want to persist this cache in redux.
+//       (This works currently fine for a `build`, but won't for `develop`)
+let mappedByKey
```

Comment:
So there are two options - if we prefer to do full rebuild, then this should be in schema/node-model.js (which holds other "node state" resettable stuff). Then you could pass it down the runQuery chain.

Other option is to do auto rebuilds when data updates via Redux. This means that we'll have additional redux actions to ensure index after which node updates will update the index.

Reply:
This generally makes me think of the right level of abstraction to have it on - from one hand it's all sift specific, on the other hand indexing logic might be shared between backends. node-model is "backend independent", however maybe our logic when we want indexes is not?

Reply:
Frankly we aren't going to get non redux/sift backend anytime soon and we had enough trouble trying to make loki work. Let's stick it into node-model.

Reply:
In the case of 150k pages (with a ton more nodes) the rebuild takes 350ms. That's neglible in the grand scheme of things so I don't think aggressive rebuilding is a concern right now (_I'm sorry, future self_).

Reply:
:+1: